lib/semantics/check-omp-structure.cc

```diff
@@ -82,10 +103,55 @@ void OmpStructureChecker::Enter(const parser::OpenMPLoopConstruct &x) {
   // push a context even in the error case
   OmpContext ct{dir.source};
   ompContext_.push_back(ct);
+  checkState_.justSawLoopConstruct = true;
+}
+
+void OmpStructureChecker::Enter(const parser::ExecutableConstruct &x) {
+  if (checkState_.justSawLoopConstruct) {
+    if (const auto *doDir{
+            std::get_if<common::Indirection<parser::DoConstruct>>(&x.u)}) {
+      // save associated DoConstruct pointer
+      SetContextDo(doDir->value());
+      checkState_.inOmpLoopRegion = true;
+    } else {
+      context_.Say(GetContext().directiveSource,
+          "do-loop is expected after the %s directive"_err_en_US,
+          ContextDirectiveAsFortran());
+      ompContext_.pop_back();
+    }
+    checkState_.justSawLoopConstruct = false;
+  }
+
+  const auto *endDoDir{
+      std::get_if<common::Indirection<parser::OpenMPEndLoopDirective>>(&x.u)};
+  if (!endDoDir && checkState_.allowOptionalEndDirective) {
+    // END directive is not present, loop construct is complete
+    checkState_.allowOptionalEndDirective = false;
+  }
 }
 
-void OmpStructureChecker::Leave(const parser::OpenMPLoopConstruct &) {
-  ompContext_.pop_back();
+void OmpStructureChecker::Leave(const parser::ExecutableConstruct &x) {
+  if (const auto *doDir{
+          std::get_if<common::Indirection<parser::DoConstruct>>(&x.u)}) {
+    if (checkState_.inOmpLoopRegion) {
+      if (&doDir->value() == GetContext().doDirective) {
+        checkState_.inOmpLoopRegion = false;
+        checkState_.allowOptionalEndDirective = true;
+        // pop at the end of associated do-loop
+        ompContext_.pop_back();
+      }
+    }
+  }
+}
+
+void OmpStructureChecker::Enter(const parser::OpenMPEndLoopDirective &x) {
```

Comment:
Should you also check that `OpenMPEndLoopDirective` macthes the `OmpLoopDirective` or will this be part of a future PR ?
Currently, the following program is accepted which seems wrong:
```
  N = 1024
  !$omp parallel do
  do i = 1, N
  enddo
  !$omp end simd
end
```

Reply:
Yes! I forgot to mention, directive matching will be in a separate PR. Thanks!